File: threads/src/main/java/im/threads/internal/transport/HistoryParser.java
package im.threads.internal.transport;

import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.google.gson.JsonSyntaxException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import im.threads.R;
import im.threads.internal.Config;
import im.threads.internal.formatters.ChatItemType;
import im.threads.internal.model.Attachment;
import im.threads.internal.model.ChatItem;
import im.threads.internal.model.ConsultConnectionMessage;
import im.threads.internal.model.ConsultPhrase;
import im.threads.internal.model.FileDescription;
import im.threads.internal.model.HistoryResponse;
import im.threads.internal.model.MessageFromHistory;
import im.threads.internal.model.MessageState;
import im.threads.internal.model.Operator;
import im.threads.internal.model.Optional;
import im.threads.internal.model.QuestionDTO;
import im.threads.internal.model.Quote;
import im.threads.internal.model.RequestResolveThread;
import im.threads.internal.model.SimpleSystemMessage;
import im.threads.internal.model.Survey;
import im.threads.internal.model.UserPhrase;
import im.threads.internal.utils.DateHelper;
import im.threads.internal.utils.ThreadsLogger;

public final class HistoryParser {
    private static final String TAG = "HistoryParser ";

    private HistoryParser() {
    }

    @NonNull
    public static List<ChatItem> getChatItems(HistoryResponse response) {
        List<ChatItem> list = new ArrayList<>();
        if (response != null) {
            List<MessageFromHistory> responseList = response.getMessages();
            if (responseList != null) {
                list = getChatItems(responseList);
                HistoryLoader.setupLastItemIdFromHistory(responseList);
            }
        }
        return list;
    }

    private static List<ChatItem> getChatItems(final List<MessageFromHistory> messages) {
        final List<ChatItem> out = new ArrayList<>();
        try {
            for (final MessageFromHistory message : messages) {
                if (message == null) {
                    continue;
                }
                final String uuid = message.getUuid();
                final String providerId = message.getProviderId();
                final List<String> providerIds = message.getProviderIds();
                final long timeStamp = message.getTimeStamp();
                final Operator operator = message.getOperator();
                String name = null;
                String photoUrl = null;
                String operatorId = null;
                boolean sex = false;
                String orgUnit = null;
                if (operator != null) {
                    name = operator.getAliasOrName();
                    photoUrl = !TextUtils.isEmpty(operator.getPhotoUrl()) ? operator.getPhotoUrl() : null;
                    operatorId = operator.getId() != null ? String.valueOf(operator.getId()) : null;
                    sex = Operator.Gender.MALE.equals(operator.getGender());
                    orgUnit = operator.getOrgUnit();
                }
                ChatItemType type = ChatItemType.fromString(message.getType());
                switch (type) {
                    case THREAD_ENQUEUED:
                    case AVERAGE_WAIT_TIME:
                    case PARTING_AFTER_SURVEY:
                    case THREAD_CLOSED:
                    case THREAD_WILL_BE_REASSIGNED:
                    case THREAD_IN_PROGRESS:
                        out.add(getSystemMessageFromHistory(message));
                        break;
                    case OPERATOR_JOINED:
                    case OPERATOR_LEFT:
                        out.add(new ConsultConnectionMessage(uuid, providerId, providerIds, operatorId, message.getType(), name, sex, timeStamp, photoUrl, null, null, orgUnit, message.isDisplay(), message.getText(), message.getThreadId()));
                        break;
                    case SURVEY:
                        Survey survey = getSurveyFromJsonString(message.getText());
                        if (survey != null) {
                            survey.setRead(message.isRead());
                            survey.setPhraseTimeStamp(message.getTimeStamp());
                            for (final QuestionDTO questionDTO : survey.getQuestions()) {
                                questionDTO.setPhraseTimeStamp(message.getTimeStamp());
                            }
                            out.add(survey);
                        }
                        break;
                    case SURVEY_QUESTION_ANSWER:
                        out.add(getCompletedSurveyFromHistory(message));
                        break;
                    case REQUEST_CLOSE_THREAD:
                        out.add(new RequestResolveThread(uuid, message.getHideAfter(), timeStamp, message.getThreadId(), message.isRead()));
                        break;
                    default:
                        final String phraseText = message.getText();
                        final FileDescription fileDescription = message.getAttachments() != null ? fileDescriptionFromList(message.getAttachments()) : null;
                        if (fileDescription != null) {
                            fileDescription.setFrom(name);
                            fileDescription.setTimeStamp(timeStamp);
                        }
                        final Quote quote = message.getQuotes() != null ? quoteFromList(message.getQuotes()) : null;
                        if (quote != null && quote.getFileDescription() != null)
                            quote.getFileDescription().setTimeStamp(timeStamp);
                        if (message.getOperator() != null) {
                            out.add(new ConsultPhrase(uuid, providerId, providerIds, fileDescription, quote, name, phraseText, message.getFormattedText(), timeStamp,
                                    operatorId, photoUrl, message.isRead(), null, false, message.getThreadId(), message.getQuickReplies()));
                        } else {
                            if (fileDescription != null) {
                                fileDescription.setFrom(Config.instance.context.getString(R.string.threads_I));
                            }
                            MessageState sentState = message.isRead() ? MessageState.STATE_WAS_READ : MessageState.STATE_SENT;
                            out.add(new UserPhrase(uuid, providerId, providerIds, phraseText, quote, timeStamp, fileDescription, sentState, message.getThreadId()));
                        }
                }
            }
            Collections.sort(out, (ci1, ci2) -> Long.compare(ci1.getTimeStamp(), ci2.getTimeStamp()));
        } catch (final Exception e) {
            ThreadsLogger.e(TAG, "error while formatting: " + messages, e);
        }
        return out;
    }

    private static Survey getSurveyFromJsonString(@NonNull String text) {
        try {
            Survey survey = Config.instance.gson.fromJson(text, Survey.class);
            final long time = new Date().getTime();
            survey.setPhraseTimeStamp(time);
            survey.setSentState(MessageState.STATE_NOT_SENT);
            survey.setDisplayMessage(true);
            for (final QuestionDTO questionDTO : survey.getQuestions()) {
                questionDTO.setPhraseTimeStamp(time);
            }
            return survey;
        } catch (final JsonSyntaxException e) {
            ThreadsLogger.e(TAG, "getSurveyFromJsonString", e);
            return null;
        }
    }

    private static Survey getCompletedSurveyFromHistory(MessageFromHistory message) {
        Survey survey = new Survey(message.getUuid(), message.getSendingId(), message.getTimeStamp(), MessageState.STATE_WAS_READ, message.isRead(), message.isDisplay());
        QuestionDTO question = new QuestionDTO();
        question.setId(message.getQuestionId());
        question.setPhraseTimeStamp(message.getTimeStamp());
        question.setText(message.getText());
        question.setRate(message.getRate());
        question.setScale(message.getScale());
        question.setSendingId(message.getSendingId());
        question.setSimple(message.isSimple());
        survey.setQuestions(Collections.singletonList(question));
        return survey;
    }

    private static SimpleSystemMessage getSystemMessageFromHistory(MessageFromHistory message) {
        return new SimpleSystemMessage(message.getUuid(), message.getType(), message.getTimeStamp(), message.getText(), message.getThreadId());
    }

    private static Quote quoteFromList(final List<MessageFromHistory> quotes) {
        Quote quote = null;
        if (quotes.size() > 0 && quotes.get(0) != null) {
            MessageFromHistory quoteFromHistory = quotes.get(0);
            FileDescription quoteFileDescription = null;
            String quoteString = null;
            String authorName;
            String receivedDateString = quoteFromHistory.getReceivedDate();
            long timestamp = receivedDateString == null || receivedDateString.isEmpty() ? System.currentTimeMillis() : DateHelper.getMessageTimestampFromDateString(receivedDateString);
            if (quoteFromHistory.getText() != null) {
                quoteString = quoteFromHistory.getText();
            }
            if ((quoteFromHistory.getAttachments() != null)
                    && (quoteFromHistory.getAttachments().size() > 0
                    && (quoteFromHistory.getAttachments().get(0).getResult() != null))) {
                quoteFileDescription = fileDescriptionFromList(quoteFromHistory.getAttachments());
            }
            if (quoteFromHistory.getOperator() != null) {
                authorName = quoteFromHistory.getOperator().getAliasOrName();
            } else {
                authorName = Config.instance.context.getString(R.string.threads_I);
            }
            if (quoteString != null || quoteFileDescription != null) {
                quote = new Quote(quoteFromHistory.getUuid(), authorName, quoteString, quoteFileDescription, timestamp);
            }
            if (quoteFileDescription != null) {
                quoteFileDescription.setFrom(authorName);
            }
        }
        return quote;
    }

    private static FileDescription fileDescriptionFromList(final List<Attachment> attachments) {
        FileDescription fileDescription = null;
        if (attachments.size() > 0) {
            Attachment attachment = attachments.get(0);
            if (attachment != null) {
                String header = null;
                String mimeType = null;
                long size = 0;
                Optional metaData = attachment.getOptional();
                if (metaData != null) {
                    header = metaData.getName();
                    size = metaData.getSize();
                    mimeType = metaData.getType();
                }
                fileDescription = new FileDescription(
                        null,
                        null,
                        size,
                        0
                );
                fileDescription.setDownloadPath(attachment.getResult());
                fileDescription.setIncomingName(header);
                fileDescription.setMimeType(mimeType);
                fileDescription.setSelfie(attachment.isSelfie());
            }
        }
        return fileDescription;
    }
}

File: src/main/java/im/threads/business/transport/HistoryParser.kt
package im.threads.business.transport

import android.text.TextUtils
import im.threads.R
import im.threads.business.config.BaseConfig
import im.threads.business.formatters.ChatItemType
import im.threads.business.formatters.ChatItemType.Companion.fromString
import im.threads.business.formatters.SpeechStatus
import im.threads.business.logger.LoggerEdna.error
import im.threads.business.models.Attachment
import im.threads.business.models.ChatItem
import im.threads.business.models.ConsultConnectionMessage
import im.threads.business.models.ConsultPhrase
import im.threads.business.models.ConsultRole
import im.threads.business.models.FileDescription
import im.threads.business.models.MessageFromHistory
import im.threads.business.models.MessageStatus
import im.threads.business.models.Operator
import im.threads.business.models.QuestionDTO
import im.threads.business.models.Quote
import im.threads.business.models.RequestResolveThread
import im.threads.business.models.SimpleSystemMessage
import im.threads.business.models.Survey
import im.threads.business.models.UserPhrase
import im.threads.business.models.enums.ModificationStateEnum
import im.threads.business.rest.models.HistoryResponse
import im.threads.business.serviceLocator.core.inject
import im.threads.business.utils.DateHelper
import java.util.ArrayList

object HistoryParser {
    private val historyLoader: HistoryLoader by inject()

    fun getChatItems(response: HistoryResponse?): List<ChatItem> {
        var list: List<ChatItem> = ArrayList()
        if (response != null) {
            val responseList = response.messages
            list = getChatItems(responseList)
            historyLoader.setupLastItemIdFromHistory(responseList)
        }
        return list
    }

    private fun getChatItems(messages: List<MessageFromHistory?>): List<ChatItem> {
        val out: MutableList<ChatItem> = ArrayList()
        try {
            for (message in messages) {
                if (message == null) {
                    continue
                }
                val uuid = message.uuid
                val timeStamp: Long = message.timeStamp
                val operator = message.operator
                var name: String? = null
                var photoUrl: String? = null
                var operatorId: String? = null
                var sex = false
                var orgUnit: String? = null
                var role: String? = null
                if (operator != null) {
                    name = operator.aliasOrName
                    photoUrl = if (!TextUtils.isEmpty(operator.photoUrl)) operator.photoUrl else null
                    operatorId = if (operator.id != null) operator.id.toString() else null
                    sex = Operator.Gender.MALE == operator.gender
                    orgUnit = operator.orgUnit
                    role = operator.role
                }
                when (fromString(message.type ?: "")) {
                    ChatItemType.THREAD_ENQUEUED, ChatItemType.AVERAGE_WAIT_TIME, ChatItemType.PARTING_AFTER_SURVEY,
                    ChatItemType.THREAD_CLOSED, ChatItemType.THREAD_WILL_BE_REASSIGNED, ChatItemType.CLIENT_BLOCKED,
                    ChatItemType.THREAD_IN_PROGRESS -> out.add(
                        getSystemMessageFromHistory(message)
                    )
                    ChatItemType.OPERATOR_JOINED, ChatItemType.OPERATOR_LEFT -> out.add(
                        ConsultConnectionMessage(
                            uuid, operatorId,
                            message.type, name, sex, timeStamp, photoUrl,
                            null, null, orgUnit, role, message.isDisplay,
                            message.text, message.threadId ?: 0L
                        )
                    )
                    ChatItemType.SURVEY -> {
                        val survey = message.content
                        if (survey != null) {
                            survey.isRead = message.read
                            survey.timeStamp = message.timeStamp
                            survey.sentState = MessageStatus.FAILED
                            survey.isDisplayMessage = true
                            survey.questions?.indices?.forEach { index ->
                                survey.questions!![index].phraseTimeStamp = message.timeStamp
                            }
                            out.add(survey)
                        }
                    }
                    ChatItemType.SURVEY_QUESTION_ANSWER -> out.add(getCompletedSurveyFromHistory(message))
                    ChatItemType.REQUEST_CLOSE_THREAD -> out.add(
                        RequestResolveThread(
                            uuid,
                            message.hideAfter ?: 0,
                            timeStamp,
                            message.threadId ?: 0,
                            message.read
                        )
                    )
                    else -> {
                        var phraseText: String? = ""
                        if (message.text != null) {
                            phraseText = message.text
                        } else if (message.speechText != null) {
                            phraseText = message.speechText
                        }
                        val fileDescription = message.attachments?.let { fileDescriptionFromList(it) }
                        if (fileDescription != null) {
                            fileDescription.from = name
                            fileDescription.timeStamp = timeStamp
                        }
                        val quote = message.quotes?.let { quoteFromList(it) }
                        quote?.fileDescription?.timeStamp = timeStamp
                        if (message.operator != null ||
                            message.modified == ModificationStateEnum.DELETED ||
                            message.modified == ModificationStateEnum.EDITED
                        ) {
                            out.add(
                                ConsultPhrase(
                                    uuid,
                                    fileDescription,
                                    message.modified,
                                    quote,
                                    name,
                                    phraseText,
                                    message.formattedText,
                                    timeStamp,
                                    operatorId,
                                    photoUrl,
                                    message.read,
                                    message.operator?.status,
                                    false,
                                    message.threadId,
                                    message.quickReplies,
                                    message.settings?.isBlockInput,
                                    SpeechStatus.fromString(message.speechStatus),
                                    ConsultRole.consultRoleFromString(role)
                                ).apply {
                                    errorMock = message.errorMock
                                }
                            )
                        } else {
                            if (fileDescription != null) {
                                fileDescription.from = BaseConfig.getInstance().context.getString(R.string.ecc_I)
                            }
                            val sentState = if (message.errorMock == true) {
                                MessageStatus.FAILED
                            } else if (message.read) {
                                MessageStatus.READ
                            } else {
                                MessageStatus.SENT
                            }
                            out.add(
                                UserPhrase(
                                    uuid,
                                    phraseText,
                                    quote,
                                    timeStamp,
                                    fileDescription,
                                    message.modified,
                                    sentState,
                                    message.threadId
                                ).apply {
                                    errorMock = message.errorMock
                                    isRead = message.read
                                }
                            )
                        }
                    }
                }
            }
            out.sortWith { ci1: ChatItem, ci2: ChatItem -> ci1.timeStamp.compareTo(ci2.timeStamp) }
        } catch (e: Exception) {
            error("error while formatting: $messages", e)
        }
        return out
    }

    private fun getCompletedSurveyFromHistory(message: MessageFromHistory): Survey {
        val survey = Survey(
            message.uuid,
            message.sendingId ?: 0,
            message.timeStamp,
            MessageStatus.READ,
            message.read,
            message.isDisplay
        )
        val question = QuestionDTO()
        question.id = message.questionId ?: 0
        question.phraseTimeStamp = message.timeStamp
        question.text = message.text
        question.rate = message.rate
        question.scale = message.scale ?: 0
        question.sendingId = message.sendingId ?: 0
        question.simple = message.isSimple
        survey.questions = arrayListOf(question)
        return survey
    }

    private fun getSystemMessageFromHistory(message: MessageFromHistory): SimpleSystemMessage {
        return SimpleSystemMessage(message.uuid, message.type, message.timeStamp, message.text, message.threadId ?: 0)
    }

    private fun quoteFromList(quotes: List<MessageFromHistory?>): Quote? {
        var quote: Quote? = null
        if (quotes.isNotEmpty() && quotes[0] != null) {
            val quoteFromHistory = quotes[0]
            var quoteFileDescription: FileDescription? = null
            var quoteString: String? = null
            val receivedDateString = quoteFromHistory!!.receivedDate
            val timestamp =
                if (receivedDateString == null || receivedDateString.isEmpty()) {
                    System.currentTimeMillis()
                } else {
                    DateHelper.getMessageTimestampFromDateString(
                        receivedDateString
                    )
                }
            if (quoteFromHistory.text != null) {
                quoteString = quoteFromHistory.text
            }
            if (quoteFromHistory.attachments != null && quoteFromHistory.attachments!!.isNotEmpty() &&
                quoteFromHistory.attachments!![0].result != null
            ) {
                quoteFileDescription = fileDescriptionFromList(quoteFromHistory.attachments!!)
            }
            val authorName = quoteFromHistory.operator?.aliasOrName ?: BaseConfig.getInstance().context.getString(R.string.ecc_I)

            if (quoteString != null || quoteFileDescription != null) {
                quote = Quote(quoteFromHistory.uuid, authorName, quoteString, quoteFileDescription, timestamp)
            }
            if (quoteFileDescription != null) {
                quoteFileDescription.from = authorName
            }
        }
        return quote
    }

    private fun fileDescriptionFromList(attachments: List<Attachment?>): FileDescription? {
        var fileDescription: FileDescription? = null
        if (attachments.isNotEmpty()) {
            val attachment = attachments[0]
            if (attachment != null) {
                val incomingName = attachment.name
                val mimeType = attachment.type
                var size: Long = 0
                val metaData = attachment.optional
                if (metaData != null) {
                    size = metaData.size ?: 0
                }
                fileDescription = FileDescription(
                    null,
                    null,
                    size,
                    0
                )
                fileDescription.downloadPath = attachment.result
                fileDescription.incomingName = incomingName
                fileDescription.mimeType = mimeType
                fileDescription.state = attachment.state
                if (attachment.errorCode != null) {
                    fileDescription.errorCode = attachment.getErrorCodeState()
                    fileDescription.errorMessage = attachment.errorMessage
                }
            }
        }
        return fileDescription
    }
}
